File: code/jvm/src/main/kotlin/pt/isel/daw/gomoku/http/media/siren/SubEntity.kt
package pt.isel.daw.gomoku.http.media.siren

import com.fasterxml.jackson.annotation.JsonProperty
import java.net.URI

/**
 * Represents a sub-entity of an entity
 */
sealed class SubEntity {

    /**
     * Represents a link that represent a navigational transition
     * @property rel the relation of the link
     * @property href the URI of the link
     */
    data class EmbeddedLink(
        val rel: List<String>,
        val href: URI
    ) : SubEntity()

    /**
     * Represents a sub-entity that is embedded in the entity
     * @property rel the relation of the sub-entity
     * @property clazz the classes of the sub-entity (optional)
     * @property properties the properties of the sub-entity (optional)
     * @property links the links of the sub-entity (optional)
     * @property actions the actions of the sub-entity (optional)
     * @property entities the sub-entities of the sub-entity (optional)
     */
    data class EmbeddedRepresentation<T>(
        val rel: List<String>,
        @get:JsonProperty("class")
        override val clazz: List<String>? = null,
        override val properties: T? = null,
        override val links: List<Link>? = null,
        override val actions: List<Action>? = null,
        override val entities: List<SubEntity>? = null
    ) : Entity<T>, SubEntity()
}

File: code/js/src/services/media/siren/SubEntity.ts
import { Link } from './Link';
import { Action } from './Action';
import { Entity } from './Entity';

export abstract class SubEntity {}

/**
 * Represents a link that represents a navigational transition
 * @param rel The relation of the link
 * @param href The URI of the link
 */
export class EmbeddedLink extends SubEntity {
  constructor(
    public rel: string[],
    public href: string
  ) {
    super();
  }
}

/**
 * Represents a sub-entity that is embedded in an entity
 * @param rel The relation of the sub-entity
 * @param classes The classes of the sub-entity (optional)
 * @param properties The properties of the sub-entity (optional)
 * @param links The links of the sub-entity (optional)
 * @param actions The actions of the sub-entity (optional)
 * @param entities The sub-entities of the sub-entity (optional)
 */
export class EmbeddedRepresentation<T> extends SubEntity implements Entity<T> {
  constructor(
    public rel: string[],
    public classes?: string[],
    public properties?: T,
    public links?: Link[],
    public actions?: Action[],
    public entities?: SubEntity[]
  ) {
    super();
  }
}
